Fix persistent volumes.

jsafrane opened the pull request:
Check Spec.PersistentVolumeSource in NFS, RBD, Gluster and iSCSI volume
plugins.

Without the patch, kubelet catches SIGSEGV in `nfsPlugin.newBuilderInternal()`:

```
    return &nfs{
        volName:    spec.Name,
        server:     spec.VolumeSource.NFS.Server,
        exportPath: spec.VolumeSource.NFS.Path,
        readOnly:   spec.VolumeSource.NFS.ReadOnly,
```

`spec.VolumeSource` is `nil` there, the volume is specified in `spec.PersistentVolumeSource`.

```
E0601 14:10:49.822245   27019 util.go:69] Recovered from panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/home/jsafrane/project/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/util.go:63
/home/jsafrane/project/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/util.go:54
/usr/lib/golang/src/runtime/asm_amd64.s:401
/usr/lib/golang/src/runtime/panic.go:387
/usr/lib/golang/src/runtime/panic.go:42
/usr/lib/golang/src/runtime/sigpanic_unix.go:26
/home/jsafrane/project/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/nfs/nfs.go:78
/home/jsafrane/project/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/nfs/nfs.go:67
/home/jsafrane/project/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volumes.go:92
/home/jsafrane/project/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volumes.go:60
/home/jsafrane/project/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/persistent_claim/persistent_claim.go:81
/home/jsafrane/project/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volumes.go:92
/home/jsafrane/project/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volumes.go:112
/home/jsafrane/project/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/kubelet.go:1120
/home/jsafrane/project/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/kubelet.go:324
/home/jsafrane/project/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/pod_workers.go:106
/home/jsafrane/project/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/pod_workers.go:115
/home/jsafrane/project/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/pod_workers.go:136
/usr/lib/golang/src/runtime/asm_amd64.s:2232
```

jsafrane commented:
cc: @markturansky

markturansky commented:
Thanks @jsafrane.  Looks good.  Can you include a unit test that exposes the bug?  Fixes look good.

jsafrane committed:
Fix persistent volumes.

Check Spec.PersistentVolumeSource in NFS, RBD, Gluster and iSCSI volume
plugins.

jsafrane commented:
Now with unit tests.

thockin commented:
Change is good but CI fails.

# github.com/GoogleCloudPlatform/kubernetes/pkg/volume/rbd
_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/rbd/rbd_test.go:68: use of package testing without selector
_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/rbd/rbd_test.go:68: undefined: volume.SpecT
FAIL	github.com/GoogleCloudPlatform/kubernetes/pkg/volume/rbd [build failed]

Please run tests before submitting :)

markturansky commented:
@thockin I fixed the test in #9069.   We can close this PR in favor of that one.

markturansky commented:
Jan is in the Czech republic and has already gone for the day, hence my hijacking this PR and providing the fix for the test.

jsafrane commented:
Oops, I would swear I ran the test several times... Anyway, thanks Marek for #9069.
